statsd/metrics.go

```diff
@@ -123,59 +124,123 @@ func (s *setMetric) flushUnsafe() []metric {
 type bufferedMetric struct {
 	sync.Mutex
 
+	// Kept samples (after sampling)
 	data []float64
+	// Total stored samples (after sampling)
+	storedSamples int64
+	// Total number of observed samples (before sampling). This is used to keep
+	// the sampling rate correct.
+	totalSamples int64
+
 	name string
 	// Histograms and Distributions store tags as one string since we need
 	// to compute its size multiple time when serializing.
 	tags  string
 	mtype metricType
+
+	// maxSamples is the maximum number of samples we keep in memory
+	maxSamples int64
 }
 
 func (s *bufferedMetric) sample(v float64) {
 	s.Lock()
 	defer s.Unlock()
+	s.sampleUnsafe(v)
+}
+
+func (s *bufferedMetric) sampleUnsafe(v float64) {
 	s.data = append(s.data, v)
+	s.storedSamples++
+	atomic.AddInt64(&s.totalSamples, 1)
```

Comment:
**question:** Why is `storedSamples` incremented through `++` while `totalSamples` is incremented through an atomic operation?

Reply:
Total samples needs to be incremented though an atomic because it can be accessed without the lock.

Adding a comment